[IMP] web: datetime_picker active text contrast

chgo-odoo opened the pull request:
=== ISSUE ===

If you go to Project > Task > Open a task and click on the Planned Date field to open the date picker, there is two contrast issues when hovering and clicking on an day.

- when hovering : The text become white which is not readable

- when clicking on a day : The background and the text have exactly the same color.


=== AFTER ===

We replicate the old design, using the primary color for the active item. We also prevent this items to receive a light background on hover since the render is weird and the text would not be readable anymore.


task-3329797
part of task-332626

Description of the issue/feature this PR addresses:

Current behavior before PR:

Desired behavior after PR is merged:




---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

robodoo commented:
[Pull request status dashboard](https://mergebot.odoo.com/odoo/odoo/pull/121615).

chgo-odoo commented:
Hi @stefanorigano, could you have a look at this PR please?

Thanks 👍

chgo-odoo commented:
@stefanorigano

stefanorigano commented on addons/web/static/src/core/datetime/datetime_picker.scss:
```diff
@@ -26,6 +26,10 @@
             color: var(--DateTimePicker__Today-highlight-color);
+            
+            &, &:hover{
+                background-color: map-get($theme-colors, primary);
```
The component makes use of CSS variables.
I'd prefer to keep using the same approach.

chgo-odoo replied:
If I understand correctly, we should use the action color for the date picker? But we don't have css variable for that color. Can't we use `$o-action` then? + `--bg-opacity` doesn't affect colors somehow

stefanorigano replied:
>  Can't we use $o-action then

Yes, in this case we can 👍 

> --bg-opacity doesn't affect the colors somehow

`--bg-opacity` can be used for `bg-[x]` classes only.

stefanorigano reviewed (changes requested):
Thanks @chgo-odoo .
Your changes would partially fix the issue indeed, but I think that we need a bit more.

In attach. the design of the previous component (`DateTimePicker` is indeed relatively [new](https://github.com/odoo/odoo/commit/b5794e89e1ad29e2a86c7ddaf241e3fc24654b5f)).
Please note that at that time the use of primary-buttons colors and active were the opposite compared to milk  ;)

---

<img width="539" style="border:1px solid black" alt="image" src="https://github.com/odoo/odoo/assets/8558351/6d6a8f92-54fc-4e16-8108-4d22719e47cb">

---

chgo-odoo commented:
@stefanorigano Changes done using the `$o-action` SCSS variable. I also fixed the `background-color` for the date between start/end and removed their border-radius because it looked weird when they were selected IMO.

Could you check? Thanks👍

stefanorigano reviewed (changes requested):
Thanks @chgo-odoo , I think we can improve more :)

In detail

- Ok for first & last button to be "full" `action-color`, but "intermediary" entries should use `$o-component-active-color` ,  `$o-component-active-border` and  `$o-component-active-bg`.  Similar to:
![image](https://github.com/odoo/odoo/assets/8558351/2fa19a6d-7a88-4a8c-8ea0-cf1bc5ff10c1).

- border radius should adapt creating a "pill" effect 
![image](https://github.com/odoo/odoo/assets/8558351/8c1b0d24-1c4e-4fb6-8708-15c7bba3576e)

- "today" should replicate the calendar design IMO, what do you think?
![image](https://github.com/odoo/odoo/assets/8558351/6e6b66c8-e7bd-449a-8448-61ce91ab6678)

chgo-odoo commented:
> * "today" should replicate the calendar design IMO, what do you think?

-------------------------
Thanks @stefanorigano  ! What about something like this for today's date?

<img width="715" alt="image" src="https://github.com/odoo/odoo/assets/128030743/af0f81ad-fbbf-4be2-900c-d4e51ecffc92">

chgo-odoo commented:
Good morning @stefanorigano !

It appears that all the issues have been handled by this PR :
- https://github.com/odoo/odoo/pull/121145

Of course we can still improve it by modifying the `primary` to `action` and modify the render of the `today date` but this is more like improvements. 

What's your thoughts on this? Should we rename this PR? Should we rework it to only add our color modification and the one to the today date?

stefanorigano reviewed (changes requested):
Thanks @chgo-odoo ,
I've found a clitch when hovering the current day in "one-date" mode only.
<img width="748" alt="image" src="https://github.com/odoo/odoo/assets/8558351/fbeb0bbd-81ef-4d12-94f6-d3f6ae906e5f">

In terms of design, I think that something like this would be more appropriate
<img width="313" alt="image" src="https://github.com/odoo/odoo/assets/8558351/fd44234d-6176-483b-9211-9b6a4dfb782a">
What do you think?

chgo-odoo commented:
I'll have a look at it in the afternoon !

chgo-odoo commented:
Good morning @stefanorigano, made some test yesterday and two days ago and I believe we can't achieve the exact same result with borders. Of course if you select dates from the same week it looks great, but selecting date from two different weeks looks weird.

This is due to the fact that border adds up to the upper row one. We can't fix this with a margin because it will cause a flickering effect on hover, and box-shadow doesn't work either because it would not render on the last row.


I see two options there : 


1/ No more border, We keep border on the `start` and `end` but remove the border of the date in the range : 

![image](https://github.com/odoo/odoo/assets/128030743/5449d451-20fd-418c-8c1d-f6e6b857813b)


2/ We use an inset to add the border : 

![image](https://github.com/odoo/odoo/assets/128030743/fcce83f1-907e-4311-8965-de3ab4d04948)

-------------------------------

I'm not really convinced about the first option because having a border only on the `start` and `end` feels a bit weird.

We could actually try to make some variants of the second option:
![image](https://github.com/odoo/odoo/assets/128030743/233e539a-73de-4268-8b9e-b35c6aa37df6)

stefanorigano commented:
Hello @chgo-odoo , thanks
I think we can address several design issues using a pseudo-element, eg:


```scss 
.o_datetime_picker .o_select_start:before, .o_datetime_picker .o_select_end:before {
    content: "";
    position: absolute;
    background: #b6e6e2;
    box-shadow: inset 0 0 1px 1px #009688;
    width: 100%;
    height: 100%;
    border-radius: 100%;
}
``` 

![image](https://github.com/odoo/odoo/assets/8558351/e8b597c9-f2ab-4ec4-91be-782606410132) 



Could you check? thanks

chgo-odoo commented:
> Hello @chgo-odoo , thanks I think we can address several design issues using a pseudo-element, eg:
> ![image](https://user-images.githubusercontent.com/8558351/242847275-e8b597c9-f2ab-4ec4-91be-782606410132.png)


Totally agree, this is the direction I took on my Localhost. I will continue in that direction and will ping you when there is something new for you to see !

Thanks !

chgo-odoo commented:
@stefanorigano Changes have been implemented. You can check on https://42427869-saas-16-3-all.runbot169.odoo.com

Thanks 👍

chgo-odoo commented:
Hi @stefanorigano ! 

Changes done to fit what we discussed this morning. Could you have a look please?

Thanks 👍

stefanorigano commented on addons/web/static/src/core/datetime/datetime_picker.scss:
```diff
@@ -21,80 +21,83 @@
             var(--DateTimePicker__Cell-border-radius-end)
             var(--DateTimePicker__Cell-border-radius-end)
             var(--DateTimePicker__Cell-border-radius-start);
+        position: relative;
```
Convention: position directives should be placed before border (https://www.odoo.com/documentation/master/contributing/development/coding_guidelines.html?highlight=css#properties-order)

stefanorigano commented on addons/web/static/src/core/datetime/datetime_picker.scss:
```diff
@@ -21,80 +21,83 @@
-    .o_highlighted {
-        --bg-opacity: 0.25;
+    .o_datetime_button span {
+        z-index: 2;
```
can we make use of `z-index-0` and `z-index-1` utility classes instead?

stefanorigano commented on addons/web/static/src/core/datetime/datetime_picker.scss:
```diff
@@ -21,80 +21,83 @@
+    .o_current:not(.o_select_start,.o_select_end),
+    .o_highlight_start:not(.o_selected),
+    .o_highlight_end:not(.o_selected),
+    .o_select_start, .o_select_end {
```
Since you're using one line for each selector, I suggest
 
```suggestion
    .o_select_start,
    .o_select_end {
```

stefanorigano reviewed (changes requested):
@chgo-odoo , very good, thanks, some minor remarks for you.
Next steps:
- rebase the branch, you'll get a fix adding a border to dropdown-menu and popovers. You should adapt the design accordingly.

- input fields and buttons still need some love ;)  

- FP proposes to redesign the red dot as  a "line", I think it's a good idea. 
<img width="366" alt="image" src="https://github.com/odoo/odoo/assets/8558351/759d36fe-77ab-4b50-b52c-ad887bb52015">

chgo-odoo commented:
Hi @stefanorigano 👋 thanks for the feedback !

Changes have been done !

**Here is a little changelog :**

- Added `.rounded` class to fix the white background owerflowing the border ;
- Made the top navigation buttons `.btn-light` ;
- Added `.rounded` class to the time picker input at the bottom ;
- Removed the absolute positioning of the primary button at the bottom and added a `apply` label to it ;
- Fixed the layout of that bottom section because it has a weird behaviour on mobile devices ;
- Replaced the `.overflow-hidden` of the date picker by `.overflow-auto` -> On mobile devices, the whole calendar is not shown and `.overflow-hidden` prevents the user to swipe and display the hidden date.

You can check the changes [on that runbot](https://45024414-saas-16-3-all.runbot125.odoo.com/).

What's your thoughts on this?

stefanorigano commented on addons/web/static/src/core/datetime/datetime_picker.scss:
```diff
@@ -16,85 +16,84 @@
+            bottom: $spacer * -.25;
+            border-radius: $border-radius-pill;
+            left: 50%;
+            transform: translateX(-50%);
```
I suggest to design the dash a bit more slim
![image](https://github.com/odoo/odoo/assets/8558351/12c17c03-64d4-40ea-8fd2-8e30e0149c5e)

example, compiled CSS:
```css
    content: "";
    position: absolute;
    background: #CE151F;
    width: 0.95em;
    height: 0.2em;
    bottom: -3px;
    border-radius: 41.9rem;
    left: 50%;
    transform: translateX(-50%);
```

stefanorigano commented on addons/web/static/src/core/datetime/datetime_picker.scss:
```diff
@@ -121,9 +120,20 @@
+        }
+    }
+
+    .o_zoom_out,.o_date_item_picker .o_datetime_button {
```
I'd use `btn-light`instead ;)

stefanorigano commented on addons/web/static/src/core/datetime/datetime_picker.xml:
```diff
@@ -84,29 +83,29 @@
         <!-- Requires: { unitIndex: number, unitList: [any, string][], timeValue_index: number } -->
         <select
-            class="o_time_picker_select o_input border-1"
+            class="o_time_picker_select o_input border-1 rounded"
```
I'd use a combination of `form-control form-control-sm` classes instead

stefanorigano commented on addons/web/static/src/core/datetime/datetime_picker.xml:
```diff
@@ -123,24 +122,24 @@
                 >
                     <t t-foreach="titles" t-as="title" t-key="title">
-                        <h5 class="o_header_part fw-bold m-0" t-esc="title" />
+                        <h5 class="o_header_part fw-bold m-0 w-50" t-esc="title" />
```
Since we want to customise the font-weight, I propose
```suggestion
                        <strong class="o_header_part w-50 fs-5" t-esc="title" />
```

Also, I would apply `w-50` only if:
- Screen => LG
- Both tables are rendered

...so we let the title expand when it's possible. Consider using `flex-basis-x`utility classes, to be tested  ;)

stefanorigano commented on addons/web/static/src/core/datetime/datetime_picker.xml:
```diff
@@ -123,24 +122,24 @@
                 </button>
-                <button class="o_next btn btn-secondary flex-grow-0" t-on-click="next">
+                <button class="o_next btn btn-light flex-grow-0" t-on-click="next">
                     <span
```
Convention; icons should use the `<i>` tag

stefanorigano commented on addons/web/static/src/core/datetime/datetime_picker.xml:
```diff
@@ -123,24 +122,24 @@
     <t t-name="web.DateTimePicker" owl="1">
         <div
-            class="o_datetime_picker bg-view d-flex flex-column p-3 gap-2 user-select-none"
+            class="o_datetime_picker rounded bg-view d-flex flex-column p-3 gap-2 user-select-none"
```
`rounded` and `bg-view` seem unnecessary. Colkd you check?

chgo-odoo commented on addons/web/static/src/core/datetime/datetime_picker.scss:
```diff
@@ -121,16 +120,24 @@
     }
 
+    .o_zoom_out {
+        gap: 72px;
```
Using buttons variables only :
`$btn-border-width * 4` (4px) + `$o-font-size-base * 2` (28px) + `$btn-padding-x * 4` (40px)

chgo-odoo commented on addons/web/static/src/core/datetime/datetime_picker.xml:
```diff
@@ -123,24 +122,24 @@
                 </button>
                 <button
-                    class="o_zoom_out o_datetime_button btn d-flex align-items-center justify-content-around text-truncate opacity-100"
+                    class="o_zoom_out o_datetime_button btn btn-light d-flex align-items-center justify-content-around px-0 text-truncate opacity-100"
```
`.px-0` is used to remove the padding of the size and correctly center the title on the calendar but `.opacity-100` is indeed unnecessary

chgo-odoo commented:
Good afternoon @stefanorigano 👋

Runbot is green (😮‍💨) and changes have been applied. I also replied to two of your notes (about `btn-light` and `px-0`).

Could you have a look at the changes, please?

Thanks !

stefanorigano reviewed (approved):
Good job @chgo-odoo  👍 

@pparidans , here another milk fix.
@Arcasias you may want to have a look too. The prev. design was good already we had just to "milkify" it.

Thanks 👍

chgo-odoo commented:
Hi @Arcasias  👋

That's a good catch, what I was trying to achieve is to avoid having a grey background when hovering today's date but this shouldn't affects the aesthetic of the date picker, I will fix this right now so we can move this PR forward 🚀

chgo-odoo commented:
Hi @stefanorigano, [could you check the changes](https://github.com/odoo/odoo/compare/089ef19caa44352117f4f056b525a6e91805c03b..e417d1a044c398015946cfd4a2fe39450c9535c3) made based on JUM's feedback, please ? Here are two little notes about the changes, thanks !  

- When `.o_today` is included inside the range, hovering it will not change its `background-color` ;

- We define a `background-color` when hovering `.o_select_start` & `.o_select_end`, only if they exist => Why do we need tis ? Because on hover they would have a grey background due to the `.btn-light` class, and if they have a transparent background, you would see the `before` & `after` pseudo-elements appear behind.

chgo-odoo commented:
Hi @stefanorigano 👋

Changes done according to our discussion this morning. Could you please check these two diff to see what really changed? Thanks ! 

- [First diff where we replace the `z-index` with a  `bg-transparent`](https://github.com/odoo/odoo/compare/bed860ffc908d8b01d41b6a651f86b773bdfb723..79c1230a67aa9ea4f822dbba15f81dde2a2945d3)

=> This avoids hard-coding a `background-color` in case `.o_today` is also `.o_select_start` and `.o_select_end`

- [Second diff where we specify that on `hover`, every button which is not `.o_select_start` or `.o_select_end` or `.o_today` will be `$o-component-active-bg` on `hover`](https://github.com/odoo/odoo/compare/79c1230a67aa9ea4f822dbba15f81dde2a2945d3..4436db2d7d5afefd3abf73a09e295cf7e45f68c7)

chgo-odoo commented:
Good evening @Arcasias 👋

I know you might have more important task going on but there are new changes since your last review and it would be great to have your feedback on it since you have been a huge contributor of the datetime picker !

Really appreciate if you manage to ! Many thanks !

chgo-odoo commented:
Hi @pparidans 👋

I know you have a lot of stuff going on but could you please have a look at this PR and eventually r+ ?

Thank you !

pparidans reviewed (approved):
LGTM
@robodoo r+

robodoo commented:
@chgo-odoo @pparidans staging failed: ci/runbot on e3266a16950598c32e6446d77b57b5992fa32622 (view more at https://runbot.odoo.com/runbot/build/47063984)

pparidans commented:
@robodoo r+

robodoo commented:
@chgo-odoo @pparidans staging failed: ci/runbot on 8689885dad75a673ee02a8ee10b1b51a7035766d (view more at https://runbot.odoo.com/runbot/build/47106529)

pparidans reviewed (commented):
:warning: @chgo-odoo Looks like the `.o_zoom_out` computation breaks in design-themes (something about mismatching `rem` and `px`)...
https://runbot.odoo.com/runbot/batch/1091803/build/47107586
